File: src/main/scala/jk_5/nailed/server/map/gamerule/DefaultEditableGameRules.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.map.gamerule

import jk_5.nailed.api.mappack.gamerule.EditableGameRules

import scala.collection.mutable

/**
 * No description given
 *
 * @author jk-5
 */
class DefaultEditableGameRules extends EditableGameRules {

  val rules = mutable.HashMap[String, String]()

  override def apply(key: String): String = this.rules.getOrElse(key, "")
  override def update(key: String, value: String): Unit = this.rules.put(key, value)
  override def ruleExists(key: String) = this.rules.contains(key)
  override def list: Seq[String] = this.rules.keySet.toSeq

  override def toString = s"DefaultEditableGameRules{}"
}

File: src/main/java/jk_5/nailed/server/map/gamerule/DefaultEditableGameRules.java
package jk_5.nailed.server.map.gamerule;

import jk_5.nailed.api.gamerule.DefaultGameRuleKey;
import jk_5.nailed.api.gamerule.EditableGameRule;
import jk_5.nailed.api.gamerule.EditableGameRules;
import jk_5.nailed.api.gamerule.GameRuleKey;

import javax.annotation.Nonnull;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * No description given
 *
 * @author jk-5
 */
public class DefaultEditableGameRules implements EditableGameRules {

    private final Map<GameRuleKey<?>, EditableGameRule<?>> gameRuleMap = new HashMap<GameRuleKey<?>, EditableGameRule<?>>();

    public DefaultEditableGameRules() {
        for(GameRuleKey<?> key : DefaultGameRuleKey.list()){
            //noinspection unchecked
            gameRuleMap.put(key, new DefaultEditableGameRule(key));
        }
    }

    @Nonnull
    @Override
    public <T> EditableGameRule<T> get(@Nonnull GameRuleKey<T> key) {
        //noinspection unchecked
        return (EditableGameRule<T>) gameRuleMap.get(key);
    }

    @Nonnull
    @Override
    public Collection<EditableGameRule<?>> list() {
        return Collections.unmodifiableCollection(gameRuleMap.values());
    }

    @Override
    public String toString() {
        return "DefaultEditableGameRules{" +
                "gameRuleMap=" + gameRuleMap +
                '}';
    }
}
